pulsar-common/src/main/java/com/yahoo/pulsar/common/util/collections/ConcurrentLongPairSet.java

```diff
@@ -0,0 +1,524 @@
+            sections[i] = new Section(perSectionCapacity);
+        }
+    }
+
+    public long size() {
+        long size = 0;
+        for (Section s : sections) {
+            size += s.size;
+        }
+        return size;
+    }
+
+    public long capacity() {
+        long capacity = 0;
+        for (Section s : sections) {
+            capacity += s.capacity;
+        }
+        return capacity;
+    }
+
+    public boolean isEmpty() {
+        for (Section s : sections) {
+            if (s.size != 0) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    long getUsedBucketCount() {
+        long usedBucketCount = 0;
+        for (Section s : sections) {
+            usedBucketCount += s.usedBuckets;
+        }
+        return usedBucketCount;
+    }
+
+    public boolean contains(long item1, long item2) {
+        checkBiggerEqualZero(item1);
+        long h = hash(item1, item2);
+        return getSection(h).contains(item1, item2, (int) h);
+    }
+
+    public boolean add(long item1, long item2) {
+        checkBiggerEqualZero(item1);
+        long h = hash(item1, item2);
+        return getSection(h).add(item1, item2, (int) h);
+    }
+
+    /**
+     * Remove an existing entry if found
+     *
+     * @param item1
+     * @return true if removed or false if item was not present
+     */
+    public boolean remove(long item1, long item2) {
+        checkBiggerEqualZero(item1);
+        long h = hash(item1, item2);
+        return getSection(h).remove(item1, item2, (int) h);
+    }
+
+    private final Section getSection(long hash) {
+        // Use 32 msb out of long to get the section
+        final int sectionIdx = (int) (hash >>> 32) & (sections.length - 1);
+        return sections[sectionIdx];
+    }
+
+    public void clear() {
+        for (Section s : sections) {
+            s.clear();
+        }
+    }
+
+    public void forEach(ConsumerLong processor) {
+        for (Section s : sections) {
+            s.forEach(processor);
+        }
+    }
+
+    /**
+     * Removes all of the elements of this collection that satisfy the given predicate.
+     * 
+     * @param filter
+     *            a predicate which returns {@code true} for elements to be removed
+     * @return {@code true} if any elements were removed
+     * 
+     * @return
+     */
+    public boolean removeIf(Predicate<LongPair> filter) {
+        boolean removed = false;
+        for (Section s : sections) {
+            if (s.removeIf(filter)) {
+                removed = true;
+            }
+        }
+        return removed;
+    }
+
+    /**
+     * @return a new list of all keys (makes a copy)
+     */
+    public Set<LongPair> items() {
+        Set<LongPair> items = new HashSet<>();
+        forEach(items::add);
+        return items;
+    }
+
+    /**
+     * @return a new list of keys with max provided numberOfItems (makes a copy)
+     */
+    public Set<LongPair> items(int numberOfItems) {
```

Comment:
Ok, I saw on the other PR that the deletion from the set is delayed to the point that messages are actually delivered to any consumer, so it makes sense this method here.

On the other hand, we could change the logic a little bit there : 
 * Remove from pendingMessage immediately
 * If there are no consumers available, put the message ids back into the pendingMessages